c7n/resources/cloudfront.py

```diff
@@ -65,3 +67,182 @@ def get_dimensions(self, resource):
         return [{'Name': self.model.dimension,
                  'Value': resource[self.model.id]},
                 {'Name': 'Region', 'Value': 'Global'}]
+
+
+@Distribution.action_registry.register('disable')
+class DistributionDisableAction(BaseAction):
+    """Action to disable a Distribution
+
+    :example:
+
+        .. code-block: yaml
+
+            policies:
+              - name: distribution-delete
+                resource: distribution
+                filters:
+                  - type: value
+                    key: CacheBehaviors.Items[].ViewerProtocolPolicy
+                    value: allow-all
+                    op: contains
+                actions:
+                  - type: disable
+    """
+    schema = type_schema('disable')
+    permissions = ("distribution:GetDistributionConfig",
+                   "distribution:UpdateDistribution",)
+
+    def process(self, distributions):
+        with self.executor_factory(max_workers=2) as w:
+            list(w.map(self.process_distribution, distributions))
+
+    def process_distribution(self, distribution):
+        client = local_session(
+            self.manager.session_factory).client(self.manager.get_model().service)
+        try:
+            res = client.get_distribution_config(
+                Id=distribution[self.manager.get_model().id])
+            res['DistributionConfig']['Enabled'] = False
+            res = client.update_distribution(
+                Id=distribution[self.manager.get_model().id],
+                IfMatch=res['ETag'],
+                DistributionConfig=res['DistributionConfig']
+            )
+        except Exception as e:
+            self.log.warning(
+                "Exception trying to disable Distribution: %s error: %s",
+                distribution['ARN'], e)
+            return
+
+
+@StreamingDistribution.action_registry.register('disable')
+class StreamingDistributionDisableAction(BaseAction):
+    """Action to disable a Streaming Distribution
+
+    :example:
+
+        .. code-block: yaml
+
+            policies:
+              - name: streaming-distribution-delete
+                resource: streaming-distribution
+                filters:
+                  - type: value
+                    key: S3Origin.OriginAccessIdentity
+                    value: ''
+                actions:
+                  - type: disable
+    """
+    schema = type_schema('disable')
+
+    permissions = ("streaming-distribution:GetStreamingDistributionConfig",
+                   "streaming-distribution:UpdateStreamingDistribution",)
+
+    def process(self, distributions):
+        with self.executor_factory(max_workers=2) as w:
+            list(w.map(self.process_distribution, distributions))
```

Comment:
not relevant in this context due to number of distributions typically being low, but for future ref any bulk resource work its typically better to pass a chunk to avoid the thread communication overhead.